fix(ivy): support inserting a `viewRef` that is already present

petebacondarwin opened the pull request:
When inserting a `viewRef` it is possible to not provide
an `index`, which is regarded as appending to the end of
the container.

If the `viewRef` already exists in the container, then
this results in a move. But there was a fault in the logic
that computed where to insert the `viewRef` that did not
account for the fact that the `viewRef` was already in
the container, so the insertion `index` was outside the
bounds of the array.

Fixes #33924

petebacondarwin committed:
fix(ivy): support inserting a `viewRef` that is already present

When inserting a `viewRef` it is possible to not provide
an `index`, which is regarded as appending to the end of
the container.

If the `viewRef` already exists in the container, then
this results in a move. But there was a fault in the logic
that computed where to insert the `viewRef` that did not
account for the fact that the `viewRef` was already in
the container, so the insertion `index` was outside the
bounds of the array.

Fixes #33924

pkozlowski-opensource commented on packages/core/test/render3/view_container_ref_spec.ts:
```diff
@@ -470,6 +472,50 @@ describe('ViewContainerRef', () => {
       expect(dynamicComp.foo.first.nativeElement).toEqual(fooEl as any);
     });
 
+    it('should support re-inserting components into view queries for dynamically created components',
```
@petebacondarwin at this point we are trying to avoid hand-written "generated" code so it should be a regular acceptance test. Plus, I think that we can have it simplified when putting alongside existing acceptance tests (there are supporting components / directives in place).

I've created a draft of such test here: https://github.com/angular/angular/pull/34054/files#diff-ac402b07939713de4f9b2a6ff5615059R208

pkozlowski-opensource commented on packages/core/src/render3/view_engine_compatibility.ts:
```diff
@@ -245,14 +245,16 @@ export function createContainerRef(
           // references appropriately.
-          return this.move(viewRef, adjustedIdx);
+          // Note that we "shift" -1 because the move will involve inserting
+          // one view but also removing one view.
```
Hmmm, this is interesting. Do you think that we could avoid doing any processing in the case where we "move" a view to the _same_ index? How about https://github.com/angular/angular/pull/34054/files#diff-f0a30efe7a958f4c73731e33aa32ebfaR272-R283 instead?

petebacondarwin replied:
I considered something similar to your suggestion but that would add yet more code to the bundle size...

I wasn't sure what the best trade off is. I guess the real question is how often would this happen. If we have change detection setup that constantly tries to insert a view into a container where it already exists (such as in ngx-formly), then this would indeed happen a lot.

BTW `index === undefined ? this.length - 1 : index)` is identical to `this._adjustIndex(index, -1)`

pkozlowski-opensource replied:
> BTW index === undefined ? this.length - 1 : index) is identical to this._adjustIndex(index, -1)

Right! 

> I considered something similar to your suggestion but that would add yet more code to the bundle size...

Hmm, as far as I can tell the total amount of added code would amount to `if (a !== b)` (after minification) so I would consider it as "worth it". Not married to the idea, though! 

Pushed an update to https://github.com/angular/angular/pull/34054

pkozlowski-opensource reviewed (changes requested):
@petebacondarwin ha! Thnx for looking into it, I was wondering what was going on when I saw the error / issue on GH! 

I left 2 comments in the PR, but when I was writing those comments I wanted to verify things and ended up doing another test + fix, see https://github.com/angular/angular/pull/34054...

Totally up to you to pick it up, change this PR or do whatever is right :-) In any case we want to avoid hand-written "generated" code.

petebacondarwin committed:
perf(ivy): do no work if moving a `viewRef` to the same position

Move a view only if it would end up at a different place.
Otherwise we would do unnecessary processing like DOM manipulation, query notifications etc.

Thanks to @pkozlowski-opensource for the change.

petebacondarwin commented:
I took your perf and test suggestions. Thanks @pkozlowski-opensource.

pkozlowski-opensource reviewed (approved):
LGTM

AndrewKushnir commented:
- [VE Presubmit](https://test.corp.google.com/ui#id=OCL:282580612:BASE:282580632:1574789095121:b3f17fa8)
- [Ivy Presubmit](https://test.corp.google.com/OCL:282580612:BASE:282580757:1574789142654:831d9a00)
